File: cliente-python/src/link/data_frame_codec.py
from link.hamming74 import encode_data_bits, recover_data_bits
from protocol.data_message import DataMessage


def encode_data_message(data_message: DataMessage) -> str:
    data_bits = text_to_bits(data_message.serialize())
    return encode_data_bits(data_bits)


def decode_data_message(codeword_bits: str) -> tuple[DataMessage, int]:
    recovered_data_bits = recover_data_bits(codeword_bits)
    serialized_message = bits_to_text(recovered_data_bits.data_bits)
    return DataMessage.parse(serialized_message), recovered_data_bits.corrected_block_count


def text_to_bits(text: str) -> str:
    return "".join(f"{byte:08b}" for byte in text.encode("utf-8"))


def bits_to_text(bits: str) -> str:
    if len(bits) % 8 != 0:
        raise ValueError("Los bits recuperados deben ser múltiplo de 8.")
    bytes_from_bits = bytes(int(bits[index:index + 8], 2) for index in range(0, len(bits), 8))
    return bytes_from_bits.decode("utf-8")

File: servidor-java/src/link/DataFrameCodec.java
package link;

import protocol.DataMessage;

import java.nio.charset.StandardCharsets;


public final class DataFrameCodec {
    private DataFrameCodec() {
    }

    public static String encode(DataMessage dataMessage) {
        byte[] messageBytes = dataMessage.serialize().getBytes(StandardCharsets.UTF_8);
        StringBuilder dataBits = new StringBuilder();

        for (byte messageByte : messageBytes) {
            for (int bitPosition = 7; bitPosition >= 0; bitPosition--) {
                dataBits.append((messageByte >> bitPosition) & 1);
            }
        }

        return Hamming74.encodeDataBits(dataBits.toString());
    }

    public static DecodedDataMessage decode(String protectedBits) {
        Hamming74.RecoveredDataBits recoveredBits = Hamming74.recoverDataBits(protectedBits);
        String dataBits = recoveredBits.getDataBits();

        if (dataBits.length() % 8 != 0) {
            throw new IllegalArgumentException("Los datos recuperados no completan bytes UTF-8.");
        }

        byte[] messageBytes = new byte[dataBits.length() / 8];

        for (int byteIndex = 0; byteIndex < messageBytes.length; byteIndex++) {
            int startIndex = byteIndex * 8;
            int byteValue = Integer.parseInt(dataBits.substring(startIndex, startIndex + 8), 2);
            messageBytes[byteIndex] = (byte) byteValue;
        }

        DataMessage dataMessage = DataMessage.parse(
            new String(messageBytes, StandardCharsets.UTF_8)
        );
        return new DecodedDataMessage(dataMessage, recoveredBits.getCorrectedBlockCount());
    }

    public static final class DecodedDataMessage {
        private final DataMessage dataMessage;
        private final int correctedBlockCount;

        private DecodedDataMessage(DataMessage dataMessage, int correctedBlockCount) {
            this.dataMessage = dataMessage;
            this.correctedBlockCount = correctedBlockCount;
        }

        public DataMessage getDataMessage() {
            return dataMessage;
        }

        public int getCorrectedBlockCount() {
            return correctedBlockCount;
        }
    }
}
